tsdb/engine/tsm1/encoding.go

```diff
@@ -111,26 +111,45 @@ func (a Values) Merge(b Values) Values {
 		return a
 	}
 
-	var i, j int
-	for ; i < len(a) && j < len(b); i++ {
-		av, bv := a[i].UnixNano(), b[j].UnixNano()
+	for i := 0; i < len(a) && len(b) > 0; i++ {
+		av, bv := a[i].UnixNano(), b[0].UnixNano()
+		// Value in a is greater than B, we need to merge
 		if av > bv {
-			a[i], b[j] = b[j], a[i]
+			// Save value in a
+			temp := a[i]
+
+			// Overwrite a with b
+			a[i] = b[0]
+
+			// Slide all values of b down 1
+			copy(b, b[1:])
+			b = b[:len(b)-1]
```

Comment:
Just a thought, but is it worth waiting until we know what to do with the merge, before we reduce the length of `b`?
Something like this maybe?

```go
copy(b, b[1:])

// See where.....
k := sort.Search(len(b)-1, func(i int) bool { return b[i].UnixNano() >= temp.UnixNano() })

// b is currently 1 element longer than possibly needed.
if k == len(b)-1 {
    b[k] = temp
} else if .... {
    last := b[len(b)-2]
    copy(b[k+1:], b[k:])
    b[len(b)-1] = last
    b[k] = temp
} else {
    // Reduce the length of b.
    b = b[:len(b)-1]
}
```

Just a thought, I've not tested it or anything! Could reduce some allocations here though if the logic makes sense.